### ISSUE TITLE
RemoteEndpoint unavailable, current state [CLOSING], expecting [OPEN or CONNECTED]

### ISSUE THREAD

**Issue by jmreicha** [rancher/rancher#8330]
**Rancher Versions:**
Server: 1.5.1
healthcheck: 0.2.3
ipsec: rancher/net:holder
network-services: metadata:v0.8.11/network-manager:v0.5.3
scheduler: 0.7.5

**Docker Version:** 1.12.6

**OS and where are the hosts located? (cloud, bare metal, etc):** CoreOS AWS

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)** HA setup behind ELB and connected to RDS.

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)** Cattle.

**Steps to Reproduce:**  No idea.  This is the only time I have ever seen this happen, just trying to get some ideas of what might have happened.

**Results:** 

Last night I lost 3 of my rancher agents spontaneously. I SSH'd to the server, and everything looked fine, as far as I could tell. All of the containers were still running, just weren't connected to the cluster any longer. I cleared the rancher-agent and the other containers on the host, and re-ran the rancher bootstrap script and the node rejoined the cluster.

I can't figure out why the agents randomly disconnected though, which is concerning. I found the following in the logs in the rancher-server container.

Any ideas why there would be websocket issues? The IP address for TLS handshake error is 100% an ELB address.  I have double checked that I have proxy protocol turned on.

```
time="2017-03-28T04:32:33Z" level=info msg="Handling backend connection request."
time="2017-03-28T04:32:33Z" level=info msg="Registering backend for host 5ca832d9-e949-4e41-7b70-5b653b41ba1f with session ID e07437cc-0b52-49d9-b34b-ce8bc8d245a6."
2017/03/28 05:03:26 http: TLS handshake error from 172.26.30.88:63316: EOF
2017/03/28 05:03:29 http: TLS handshake error from 172.26.30.88:63324: EOF
time="2017-03-28T05:38:31Z" level=info msg="Shutting down backend d11849cb-d511-4055-4001-30001f8e149d. Connection closed because: websocket: close 1006 unexpected EOF."
time="2017-03-28T05:38:31Z" level=info msg="Removed backend. Key: d11849cb-d511-4055-4001-30001f8e149d. Session ID f9302ce6-8634-427f-82e8-d674978b09cb ."
time="2017-03-28T05:38:44Z" level=info msg="Handling backend connection request."
time="2017-03-28T05:38:44Z" level=info msg="Registering backend for host d11849cb-d511-4055-4001-30001f8e149d with session ID 29397ff5-664a-40ce-a3b7-77d48aedc376."
2017-03-28 05:45:31,659 WARN  [:] [] [] [] [p859417998-4093] [i.c.p.a.p.s.j.WebSocketMessageWriter] Unhandled Error (closing connection) org.eclipse.jetty.websocket.api.WebSocketException: RemoteEndpoint unavailable, current state [CLOSING], expecting [OPEN or CONNECTED]
at org.eclipse.jetty.websocket.common.WebSocketSession.getRemote(WebSocketSession.java:252) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:141) ~[websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:309) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:214) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:220) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.Parser.parse(Parser.java:258) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.readParse(AbstractWebSocketConnection.java:628) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:476) [websocket-common-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540) [jetty-io-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) [jetty-util-9.2.11.v20150529.jar:9.2.11.v20150529]
at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) [jetty-util-9.2.11.v20150529.jar:9.2.11.v20150529]
at java.lang.Thread.run(Thread.java:745) [na:1.8.0_72]
```

There are more errors in the logs but they all started happening after this first error.

**Comment by jmreicha** [comment]
I'm continuing to see other odd messages in the rancher server logs.  The following messages repeat themselves every 30 minutes or so.

```
2017-03-28 08:10:50,736 ERROR [725bf9ec-02c9-4f03-b4e5-fd96b9491e55:167179] [instance:2984] [instance.start->(InstanceStart)] [] [cutorService-13] [i.c.p.process.instance.InstanceStart] Failed [2/5] to Starting for instance [2984]
2017-03-28 08:10:50,750 ERROR [e671ad9c-9be3-472e-ba92-320a2e0208f3:167184] [instance:2990->instanceHostMap:2348] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ecutorService-6] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=956]: Timeout getting IP address
2017-03-28 08:10:50,750 ERROR [e671ad9c-9be3-472e-ba92-320a2e0208f3:167184] [instance:2990] [instance.start->(InstanceStart)] [] [ecutorService-6] [i.c.p.process.instance.InstanceStart] Failed [2/5] to Starting for instance [2990]
2017-03-28 08:10:50,869 ERROR [725bf9ec-02c9-4f03-b4e5-fd96b9491e55:167179] [instance:2984->instanceHostMap:2340] [instance.start->(InstanceStart)->instancehostmap.activate] [] [cutorService-13] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=956]: Timeout getting IP address
2017-03-28 08:10:50,869 ERROR [725bf9ec-02c9-4f03-b4e5-fd96b9491e55:167179] [instance:2984] [instance.start->(InstanceStart)] [] [cutorService-13] [i.c.p.process.instance.InstanceStart] Failed [3/5] to Starting for instance [2984]
2017-03-28 08:10:50,947 ERROR [e671ad9c-9be3-472e-ba92-320a2e0208f3:167184] [instance:2990->instanceHostMap:2348] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ecutorService-6] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=956]: Timeout getting IP address
2017-03-28 08:10:50,947 ERROR [e671ad9c-9be3-472e-ba92-320a2e0208f3:167184] [instance:2990] [instance.start->(InstanceStart)] [] [ecutorService-6] [i.c.p.process.instance.InstanceStart] Failed [3/5] to Starting for instance [2990]
2017-03-28 08:10:56,006 ERROR [3ad35107-8bd4-4981-914d-3501f9c20821:168841] [instance:3016->instanceHostMap:2369] [instance.start->(InstanceStart)->instancehostmap.activate] [] [ecutorService-9] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=956]: Timeout getting IP address
2017-03-28 08:10:56,007 ERROR [3ad35107-8bd4-4981-914d-3501f9c20821:168841] [instance:3016] [instance.start->(InstanceStart)] [] [ecutorService-9] [i.c.p.process.instance.InstanceStart] Failed [1/5] to Starting for instance [3016]
2017-03-28 08:10:56,089 ERROR [d85fd7ce-17c9-47fe-91a5-6ef93ff1b452:168835] [instance:3015->instanceHostMap:2370] [instance.start->(InstanceStart)->instancehostmap.activate] [] [utorService-888] [c.p.e.p.i.DefaultProcessInstanceImpl] Agent error for [compute.instance.activate.reply;agent=956]: Timeout getting IP address
```

**Comment by jmreicha** [comment]
Restarting the rancher-server container seems to clear things up in the logs temporarily.  I've beefed up my monitoring and alerting to catch this faster next time.

I'd still like to know if there is a proper "fix" for this behavior.  Bump the JVM memory on the server or maybe an extra flag to the docker run command to restart it automatically if it consumes too many resources?

**Comment by ecliptik** [comment]
We're seeing something similar, but it appears to be CPU related instead of memory due to possible websocket problems, https://github.com/rancher/rancher/issues/8326.

Having either more resources or a way to auto-recover would be helpful.

**Comment by jmreicha** [comment]
Fixed via #8326
